pytorch_lightning/core/lightning.py

```diff
@@ -1166,16 +1166,47 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
 
         Override for your own behavior
 
+        It works with ``untoggle_optimizer`` to make sure param_requires_grad_state is properly reset.
+
         Args:
-            optimizer:
-            optimizer_idx:
+            optimizer: Current optimizer used in training_loop
+            optimizer_idx: Current optimizer idx in training_loop
         """
-        for param in self.parameters():
-            param.requires_grad = False
+        param_requires_grad_state = {}
+        # make sure current optimizer is latest to be iterated over.
+        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
+        num_optimizers = len(optimizers) - 1
+        for opt_idx, opt in enumerate(optimizers):
+            for group in opt.param_groups:
+                for param in group['params']:
+                    if num_optimizers == opt_idx:
+                        # If a param appears in 2 optimizer, revert `requires_grad_state` to before toggle
```

Comment:
I think you meant this?

```suggestion
                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle
```

But how does "appearing in 2 optimizers" relates to this?

Reply:
We have several tests with 2 optimizers where param are shared. Assuming we are currently running on optimizer_idx = 1, optimizer_idx 0 will save the param state (let's say True) and set its `requires_grad` to `False`.

However, optimizer_idx 1 was expected to see param with `requires_grad=True`.
Therefore, at last, we need to iterate back on current optimizer and if param is in state, apply it.